ivy/func_wrapper.py

```diff
@@ -35,6 +35,175 @@
 # Helpers #
 # --------#
 
+# for casting modes, order is the hierarchy
+casting_modes_dict = {
+    "uint": ["uint8", "uint16", "uint32", "uint64"],
+    "int": ["int8", "int16", "int32", "int64"],
+    "float": ["bfloat16", "float16", "float32", "float64"],
+    "complex": ["complex64", "complex128"],
+}
+
+
+def caster(dtype, intersect):
+    if hasattr(dtype, "dtype"):
+        dtype = ivy.as_ivy_dtype(dtype.dtype)
+    else:
+        dtype = ivy.as_ivy_dtype(dtype)
+    if str(dtype) in intersect:
+        # based on upcasting or downcasting do something
+        if ivy.cast_data_types():
+            # all casting types is enabled
+            # check cross_casting
+            dtype = cross_caster(intersect)
+            if dtype:
+                return dtype
+            # check upcasting
+            dtype = upcaster(dtype)
+            if dtype:
+                return dtype
+            # check downcasting
+            dtype = downcaster(dtype)
+            if dtype:
+                return dtype
+        elif ivy.crosscast_dtypes:
+            # check cross_casting
+            dtype = cross_caster(intersect)
+            if dtype:
+                return dtype
+        elif ivy.upcast_dtypes:
+            # check upcasting
+            dtype = upcaster(dtype, intersect)
+            if dtype:
+                return dtype
+        elif ivy.downcast_dtypes:
+            # check downcasting
+            dtype = downcaster(dtype, intersect)
+            if dtype:
+                return dtype
+
+
+def upcaster(dtype, intersect):
+    # upcasting is enabled, we upcast to the highest
+    if "uint" in str(dtype):
```

Comment:
I think we can just separate out the number and the string from the dtype, so that we could reduce some of the redundancy (we wouldn't need to check if it's a `uint`, `int` or `float`)